Various typing improvements

harahu opened the pull request:
## 📚 Context

_Improve type annotations for streamlit's public API_

- What kind of change does this PR introduce?

  - [X] Other, please describe: Type annotations

## 🧠 Description of Changes

- _Relax `str` arguments for markdown methods to `SupportsStr` in line with actual behavior. The reasoning for this has been discussed in past PRs: https://github.com/streamlit/streamlit/pull/4808#discussion_r887411102_
- _An initial pass at adding types for `st.slider`. For now I've only defined the return type as the union of all possible return values. Having the return type be variable with regards to input would be more helpful, but also requires a lot more effort. I'll get back to this later._
- _Add overload definitions to selectbox, inspired by #4926_

---

**Contribution License Agreement**

By submitting this pull request you agree that all contributions to this project are made under the Apache 2.0 license.

harahu committed:
Relax typing for markdown somewhat and add initial types for slider

harahu committed:
Add overload definition for selectbox

harahu committed:
Fix test

vdonato commented on lib/streamlit/elements/markdown.py:
```diff
@@ -17,6 +17,7 @@
 from streamlit import type_util
 from streamlit.proto.Markdown_pb2 import Markdown as MarkdownProto
 from .utils import clean_text
+from ..type_util import SupportsStr
```
heads up that this will require a small change as I recently got rid of `elements.utils.type_util` in favor of an identical function in `string_util` in #5121

harahu replied:
Thanks for the heads up :)

vdonato commented on lib/streamlit/util.py:
```diff
@@ -130,7 +134,7 @@ def index_(iterable, x) -> int:
     for i, value in enumerate(iterable):
         # https://stackoverflow.com/questions/588004/is-floating-point-math-broken
         # https://github.com/streamlit/streamlit/issues/4663
```
nit: not really part of your change, but I'd remove these comments as I don't think either of them are particularly helpful given that needing to take floating point error into account shouldn't be anything surprising

harahu replied:
Comments removed.

vdonato commented on lib/streamlit/elements/selectbox.py:
```diff
@@ -13,9 +13,9 @@
 
 from textwrap import dedent
-from typing import Any, Callable, Optional, cast
+from typing import TYPE_CHECKING, Any, Callable, Optional, Sequence, TypeVar, cast
+from typing import overload
```
nit: these two import statements can be merged

sfc-gh-kbregula replied:
I hope that when we introduce isort, such problems will be detected and fixed automatically.

vdonato replied:
Yeah, agreed. It's been suggested previously, and I believe no one is in opposition to adding it as part of our build process.

My last attempt at introducing it unfortunately didn't work out, though, as I only had a few hours of free time at the time and ran into some issues that I wasn't able to quickly resolve

harahu replied:
isort would be helpful. Might look into it if I feel inspired one day. For now, the imports are merged.

vdonato commented on lib/streamlit/elements/selectbox.py:
```diff
@@ -30,8 +30,47 @@
+
 
 class SelectboxMixin:
+    @overload
+    def selectbox(
+        self,
+        label: str,
+        options: Sequence[T],
+        index: int = 0,
+        format_func: Callable[[Any], Any] = str,
+        key: Optional[Key] = None,
+        help: Optional[str] = None,
+        on_change: Optional[WidgetCallback] = None,
+        args: Optional[WidgetArgs] = None,
+        kwargs: Optional[WidgetKwargs] = None,
+        *,  # keyword-only arguments:
+        disabled: bool = False,
+    ) -> Optional[T]:
+        ...
+
+    @overload
+    def selectbox(
+        self,
+        label: str,
+        options: OptionSequence,
+        index: int = 0,
+        format_func: Callable[[Any], Any] = str,
+        key: Optional[Key] = None,
+        help: Optional[str] = None,
+        on_change: Optional[WidgetCallback] = None,
+        args: Optional[WidgetArgs] = None,
+        kwargs: Optional[WidgetKwargs] = None,
+        *,  # keyword-only arguments:
+        disabled: bool = False,
+    ) -> Optional[Any]:
+        ...
```
I'm guessing the difference between the two overloads is to allow the type checker to be able to deduce the return type of a `selectbox` in the case where `options` is a regular `Sequence` vs one of the `OptionsSequence` types that can't be parametrized? Would recommend adding a clarifying comment about this so that it's a bit more obvious.

vdonato replied:
Additionally, I think with these overloads, it's now possible to remove `Sequence[Any]` from the `OptionsSequence` type as that case should be captured by the first case.

harahu replied:
> I'm guessing the difference between the two overloads is to allow the type checker to be able to deduce the return type of a selectbox in the case where options is a regular Sequence vs one of the OptionsSequence types that can't be parametrized? Would recommend adding a clarifying comment about this so that it's a bit more obvious.

Your understanding is on point. Will add a comment.

harahu replied:
I took a few steps back, and I think I managed to gracefully turn OptionSequence into a generic type in such a way that the overload isn't needed anymore. I think this is a cleaner solution.

vdonato commented on lib/streamlit/elements/selectbox.py:
```diff
@@ -45,7 +84,7 @@ def selectbox(
         *,  # keyword-only arguments:
         disabled: bool = False,
-    ) -> Any:
+    ) -> Optional[Any]:
```
Could this (along with the other `Optional[Any]`s) just be `Any`?

Although I don't have a strong opinion on this either way since there is merit to having these be consistent with the `Optional[T]` above.

harahu replied:
For return types, the benefit of spelling out all of the types you know you can encounter, in addition to `Any`, is that the developer has to deal with these specifically. This is different from parameter types where, indeed, `Optional[Any]` is equivalent to `Any`.

Here's a simplified example that illustrates this point: https://mypy-play.net/?mypy=latest&python=3.10&gist=8afc339c79f6377558e19a4f3c6491d3

vdonato commented on lib/streamlit/elements/markdown.py:
```diff
@@ -17,6 +17,7 @@
 from streamlit import type_util
 from streamlit.proto.Markdown_pb2 import Markdown as MarkdownProto
 from .utils import clean_text
```
heads up that this will need a small change as I removed `elements.utils.clean_text` in favor of `string_util.clean_text` in #5121

harahu replied:
Addressed.

vdonato commented on lib/streamlit/elements/markdown.py:
```diff
@@ -17,6 +17,7 @@
 from streamlit import type_util
 from streamlit.proto.Markdown_pb2 import Markdown as MarkdownProto
 from .utils import clean_text
+from ..type_util import SupportsStr
```
nit: would slightly prefer to import from `streamlit.type_util` so that relative imports are only used for imports in the same directory.

(I don't think we have any strict convention for this in the codebase, but the code does for the most part follow it as far as I know)

harahu replied:
Fixed.

vdonato reviewed (commented):
This mostly LGTM!

I'd be in favor of going ahead and merging all of the changes contained in this PR aside from the ones to `st.slider`.

I'm a bit more hesistant about the slider types at the moment due to them being a bit more verbose than the other type annotations in the codebase, and also because some of the implications (like needing to cast the return value of a slider to an int to be able to pass it to `range`) feel a bit weird to me.

I'd recommend splitting out the slider changes into a separate PR as I think they'll warrant some discussion with others on the team. Otherwise, I'd be happy to approve+merge the rest of the changes in this PR.

harahu commented:
@vdonato Sorry for leaving this hanging. It's been a busy week.

I've updated `st.slider` so that it still returns `Any` for now. That should remove the need for casting the return value, which I agree is annoying. As for the parameter types I've added to `slider`, I'd suggest including those, as they offer added protection, with no downsides, IMO.

I agree that the types are a bit verbose, but I'm pretty sure this is the best that can be done for slider.

I decided to split this PR into three subset PRs, and we can move discussions to those if you feel like it.

See #5190, #5191, and #5192

harahu committed:
PR feedback

vdonato commented:
Thanks @harahu! I'll close this PR given that I believe that the three new ones capture all of the changes that were originally made here.
